packages/hub/services/gas-estimation.ts

```diff
@@ -39,24 +39,71 @@ export default class GasEstimationService {
     let provider = this.ethersProvider.getInstance(params.chainId);
     let signer = new Wallet(config.get('hubPrivateKey'));
     let scheduledPaymentModule = await this.cardpay.getSDK('ScheduledPaymentModule', provider, signer);
+
     let gas;
-    let tokenAmount = ethers.utils.parseUnits('1', 'gwei');
-    let gasTokenAmount = ethers.utils.parseUnits('0.01', 'gwei');
     switch (params.scenario) {
       case GasEstimationResultsScenarioEnum.create_safe_with_module:
         gas = (
           await scheduledPaymentModule.createSafeWithModuleAndGuardEstimation({ from: signer.address })
         ).toNumber();
         break;
+      default:
+        gas = await this.estimatePaymentExecution(params);
+    }
+
+    gasLimit = await prisma.gasEstimationResult.upsert({
+      where: {
+        chainId_scenario_tokenAddress_gasTokenAddress: {
+          chainId: params.chainId,
+          scenario: params.scenario,
+          tokenAddress: params.tokenAddress ?? '',
+          gasTokenAddress: params.gasTokenAddress ?? '',
+        },
+      },
+      create: {
+        chainId: params.chainId,
+        scenario: params.scenario,
+        tokenAddress: params.tokenAddress,
+        gasTokenAddress: params.gasTokenAddress,
+        gas: gas,
+      },
+      update: {
+        gas: gas,
+      },
+    });
+
+    return gasLimit;
+  }
+
+  private async estimatePaymentExecution(params: GasEstimationParams) {
+    if (
+      !params.tokenAddress ||
+      params.tokenAddress === '' ||
+      !params.gasTokenAddress ||
+      params.gasTokenAddress === ''
+    ) {
+      throw Error(`tokenAddress and gasTokenAddress is required in ${params.scenario}`);
+    }
+
+    let provider = this.ethersProvider.getInstance(params.chainId);
+    let networkName = convertChainIdToName(params.chainId);
+    let tokenList = this.cardpay.getConstantByNetwork('tokenList', networkName as SchedulerCapableNetworks);
+    let token = tokenList.tokens.find((t) => t.address === params.tokenAddress);
+    let gasToken = tokenList.tokens.find((t) => t.address === params.gasTokenAddress);
+    if (!token) throw Error('unknown token');
+    if (!gasToken) throw Error('unknown gas token');
+
+    //Check the decimals of tokens
+    //to avoid error when converting USD tokens
+    //because USD token's decimals is lower than native token's decimals
+    let tokenAmount = ethers.utils.parseUnits('1', token.decimals >= 18 ? 'gwei' : token.decimals);
+    let gasTokenAmount = ethers.utils.parseUnits('0.01', gasToken.decimals >= 18 ? 'gwei' : token.decimals);
```

Comment:
Would this work?

```suggestion
    let tokenAmount = ethers.utils.parseUnits('1', token.decimals);
    let gasTokenAmount = ethers.utils.parseUnits('0.01', token.decimals);
```

Reply:
There is a worst case where the user uses WETH as a transfer token and gas token. With execution gas around `140000`, the crank safe need to have WETH balance around `1500` WETH to estimate the payment execution.

`total token = transfer token amount + (transfer token amount x fee percentage) + (gas token amount x execution gas) + (gas token amount in USD)`

Reply:
The question is how much amount of each token (WETH, USDC, USDT, etc) do we want to deposit in the crank safe?

Reply:
Just had a discussion with @FadhlanR, we realized that:

1. The `gasTokenAmount` is wrongly named, it is in fact `gasTokenPrice`
2. Both `tokenAmount`and `gasTokenPrice` should be as small as possible (so that we don't have to keep a non-trivial balance of tokens in the crank) given their token decimal amount. So in decimal terms, something like `1e-decimals`.

Reply:
Sorry for the confusion I made it simple by defining `token amount = 1000 and gas price = 100` it worked for all tokens.

Reply:
Looks good to me. Can you please add a comment in the code explaining why the value of these numbers make sense? I think it's good to document these hardcoded values in case someone needs to tweak them later.

Reply:
Comment added